waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -795,36 +799,50 @@ proc validateMessage*(rlnPeer: WakuRLNRelay, msg: WakuMessage,
     # accept messages whose epoch is within +-MAX_EPOCH_GAP from the current epoch
     debug "invalid message: epoch gap exceeds a threshold", gap = gap,
         payload = string.fromBytes(msg.payload)
+    waku_rln_invalid_messages.inc(labelValues=["invalid_epoch"])
     return MessageValidationResult.Invalid
 
   if not rlnPeer.validateRoot(msg.proof.merkleRoot):
     debug "invalid message: provided root does not belong to acceptable window of roots", provided=msg.proof.merkleRoot, validRoots=rlnPeer.validMerkleRoots
+    waku_rln_invalid_messages.inc(labelValues=["invalid_root"])
     return MessageValidationResult.Invalid
 
   # verify the proof
   let
     contentTopicBytes = msg.contentTopic.toBytes
     input = concat(msg.payload, contentTopicBytes)
-    proofVerificationRes = rlnPeer.rlnInstance.proofVerify(input, msg.proof)
+
+  waku_rln_proof_verification.inc()
+  waku_rln_proof_verification_time.time:
+    let proofVerificationRes = rlnPeer.rlnInstance.proofVerify(input, msg.proof)
 
   if proofVerificationRes.isErr():
+    waku_rln_errors.inc(labelValues=["proof_verification"])
     return MessageValidationResult.Invalid
   if not proofVerificationRes.value():
     # invalid proof
     debug "invalid message: invalid proof", payload = string.fromBytes(msg.payload)
+    waku_rln_invalid_messages.inc(labelValues=["invalid_proof"])
     return MessageValidationResult.Invalid
 
   # check if double messaging has happened
   let hasDup = rlnPeer.hasDuplicate(msg)
-  if hasDup.isOk and hasDup.value == true:
-    debug "invalid message: message is a spam", payload = string.fromBytes(msg.payload)
+  if hasDup.isErr():
+    waku_rln_errors.inc(labelValues=["duplicate_check"])
+  elif hasDup.value == true:
+    debug "invalid message: message is spam", payload = string.fromBytes(msg.payload)
+    waku_rln_spam_messages.inc()
     return MessageValidationResult.Spam
 
   # insert the message to the log
   # the result of `updateLog` is discarded because message insertion is guaranteed by the implementation i.e.,
   # it will never error out
   discard rlnPeer.updateLog(msg)
   debug "message is valid", payload = string.fromBytes(msg.payload)
+  let rootLabel = $rlnPeer.validMerkleRoots.find(msg.proof.merkleRoot)
```

Comment:
That is an interesting metric, how would you think of using the `root` value in the analysis?

Reply:
It would be interesting to know how often peers are receiving messages with roots older than the most recent one.